packages/firestore/src/util/async_queue.ts

```diff
@@ -372,9 +382,13 @@ export class AsyncQueue {
    * Waits until all currently queued tasks are finished executing. Delayed
    * operations are not run.
    */
-  drain(): Promise<void> {
-    // It should still be possible to drain the queue after it is shutting down.
-    return this.enqueueEvenAfterShutdown(() => Promise.resolve());
+  async drain(): Promise<void> {
+    // Operations in the queue prior to draining may have enqueued additional
```

Comment:
I don't know if this works, but if it does you could get rid of all size accounting:

```
let currentTail : Promise<void>;
do (
  currentTail = this.tail;
  await currentTail;
} while (currentTail !== this.tail);
```

Reply:
done.